### PR TITLE
[Transform] improve irrecoverable error detection - part 2

### PR DESCRIPTION
**Pull request by hendrikmuhs**
base error handling on rest status instead of listing individual exception types

relates to #51820

### CODE REVIEW

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/bwc 
run elasticsearch-ci/default-distro

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinder.java [hunk 1]
```diff
@@ -7,19 +7,37 @@
...
 
 /**
  * Set of static utils to find the cause of a search exception.
  */
 public final class ExceptionRootCauseFinder {
 
+    /**
+     * List of rest statuses that we consider irrecoverable
+     */
+    public static Set<RestStatus> IRRECOVERABLE_REST_STATUSES = new HashSet<>(
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinder.java]: AFAIU this one could be "final".

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinder.java [hunk 2]
```diff
@@ -7,19 +7,37 @@
...
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
 
 /**
  * Set of static utils to find the cause of a search exception.
  */
 public final class ExceptionRootCauseFinder {
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinder.java]: You may find it out of scope of this PR, but this class is lacking unit tests.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinder.java [hunk 3]
```diff
@@ -67,11 +85,16 @@ public static String getDetailedMessage(Throwable t) {
...
-            if (unwrappedThrowable instanceof MapperParsingException
-                || unwrappedThrowable instanceof IllegalArgumentException
-                || unwrappedThrowable instanceof ResourceNotFoundException) {
+            if (unwrappedThrowable instanceof IllegalArgumentException) {
                 return unwrappedThrowable;
             }
+
+            if (unwrappedThrowable instanceof ElasticsearchException) {
+                ElasticsearchException elasticsearchException = (ElasticsearchException) unwrappedThrowable;
+                if (ExceptionRootCauseFinder.IRRECOVERABLE_REST_STATUSES.contains(elasticsearchException.status())) {
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinder.java]: `ExceptionRootCauseFinder.` prefix should not be needed here.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 4]
```diff
@@ -477,37 +475,53 @@ synchronized void handleFailure(Exception e) {
...
+        }
+
+        if (context.getAndIncrementFailureCount() > context.getNumFailureRetries()) {
+            failIndexer(
+                "task encountered more than "
+                    + context.getNumFailureRetries()
+                    + " failures; latest failure: "
+                    + ExceptionRootCauseFinder.getDetailedMessage(unwrappedException)
+            );
+        } else {
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java]: You could `return;` in line 511 and get rid of `else` clauses, just like you did above.

File: x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinderTests.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        );
+        bulkItemResponses.remove(7);
+
+        assertNull(ExceptionRootCauseFinder.getFirstIrrecoverableExceptionFromBulkResponses(bulkItemResponses.values()));
+    }
+
+    private static void assertFirstException(Collection<BulkItemResponse> bulkItemResponses, Class<?> expectedClass, String message) {
+        Throwable t = ExceptionRootCauseFinder.getFirstIrrecoverableExceptionFromBulkResponses(bulkItemResponses);
+        assertNotNull(t);
+        assertEquals(t.getClass(), expectedClass);
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinderTests.java]: `assertThat(..., equalTo(...))`
?

**Review by przemekwitek — APPROVED** [review]
LGTM

**Pull request merged**